### ISSUE TITLE
Params are not passed between two StackNavigators wrapped in SwitchNavigator

### ISSUE THREAD

**Issue by animaonline** [react-navigation/react-navigation#3787]
I know this has been reported earlier (https://github.com/react-navigation/react-navigation/issues/143) but there's no workaround and this is clearly a bug.

E.g.

```
const RegisterStack = StackNavigator(
  {
    RegisterStep1: RegisterStep1,
    RegisterStep2: RegisterStep2,
    RegisterStep3: RegisterStep3,
    RegisterStep4: RegisterStep4,
    RegisterStep4: RegisterStep4,
    RegisterStepFailed: RegisterStepFailed
  },
  {
    headerMode: "screen"
  }
);

const UserInfoStack = StackNavigator(
  {
    UserInfo,
  },
  {
    headerMode: "screen"
  }
);

const RootSwitchNavigator = SwitchNavigator(
  {
    Register: RegisterStack,
    UserInfo: UserInfoStack
  },
  {
    headerMode: "screen",
    initialRouteName: "Register"
  }
);
```

When I navigate to UserInfo from RegisterStep1 like so:
`this.props.navigation.navigate("UserInfo", {hello: "world });`

The params are not passed. 

Here's a snack taken from your documentation that reproduces this problem: https://snack.expo.io/r1Ih9W6tG

Notice the alert(JSON.stringify(this.props.navigation.state.params)) in HomeScreen's componentDidMount

**Comment by brentvatne** [comment]
you have two routes named UserInfo 🤔 so it's switching to that stack and passing the param in to the stack route but not to the actual screen. rename the one at top level to `UserInfoStack` or something and it shold resolve the issue

**Comment by animaonline** [comment]
I'm sorry there's a typo in my code, but the error is still reproducible, have a look at the snack, it uses the code from your documentation.

**Comment by brentvatne** [comment]
@animaonline - it's still not the expected behavior based on what is in the snack. if you want to pass the params to the route inside of a stack, navigate to that route directly.

so in the snack it navigates to `App`, but `Home` is the route in `App` that you want to receive the params. so instead you navigate to `Home`

**Comment by brentvatne** [comment]
here's an example: https://snack.expo.io/H1ZGLKZ5G

notice that I have to use `key` here so it doesn't push another route on top. this is pretty inconvenient and resolved in 2.0 via https://github.com/react-navigation/rfcs/blob/master/text/0004-less-pushy-navigate.md

**Comment by animaonline** [comment]
Ok, thank you very much, I'll check it all out :)

**Comment by brentvatne** [comment]
here's a better version of the example: https://snack.expo.io/r1RFLYW5G

**Comment by brentvatne** [comment]
tl;dr here is that if you navigate to the route for a stack rather than a specific screen in the stack, we don't pass the params though. it's unclear to me if it would be desirable for us to pass the params though in this case -- perhaps yes? but we don't know what screen we're on and what impact that might have on that screen, so it feels like a potential area to introduce bugs in your app because it's so ambiguous

**Comment by animaonline** [comment]
Yeah it all makes sense now, thanks! ;)

**Comment by Access2emma** [comment]
What if we are to navigate  from the SignInScreen to OtherScreen and with params?

**Comment by sahil290791** [comment]
@brentvatne I still think that we should allow the stack to receive params. Let's say that I want to navigate to a stack and then based on some logic want to decide which screen to show to a user, in that case I my logic decides based on the parameters sent to the stack, so in runtime I do not know which route inside stack I should navigate to.

**Comment by wzup** [comment]
@brentvatne 
> _navigate to that route directly._

Not an option, unfortunately. Because RN renders screens twice.

**Comment by rrocchi** [comment]
Just stumbled upon this. Here is how it works for me (using the code given as an example)

```
const RegisterStack = StackNavigator(
  {
    RegisterStep1: RegisterStep1,
    RegisterStep2: RegisterStep2,
    RegisterStep3: RegisterStep3,
    RegisterStep4: RegisterStep4,
    RegisterStep4: RegisterStep4,
    RegisterStepFailed: RegisterStepFailed
  },
  {
    headerMode: "screen"
  }
);

const UserInfoStack = StackNavigator(
  {
	  UserInfoScreen: UserInfo, // Give the screen a unique name as UserInfo is also the name of this stack inside the root stack
  },
  {
    headerMode: "screen"
  }
);

const RootSwitchNavigator = SwitchNavigator(
  {
    Register: RegisterStack,
    UserInfo: UserInfoStack
  },
  {
    headerMode: "screen",
    initialRouteName: "Register"
  }
);

```


Using `this.props.navigation.navigate("UserInfoScreen", {hello: "world });` from RegisterStep1, the params get passed.

Basically navigate to the screen inside the stack rather than to the stack itself.

**Comment by PowerOfM** [comment]
@rrocchi But based on your example, if you did `navigate("RegisterStep1", { hello: "world" })` from UserInfoScreen, then RegisterStep2, RegisterStep3 or any other screens will not have access to that param.

**Comment by anshumanburman** [comment]
import React, {Component} from 'react';
import {Platform, StyleSheet, Text, View} from 'react-native';
import {createAppContainer, createStackNavigator, SafeAreaView, createDrawerNavigator, createBottomTabNavigator} from 'react-navigation';
import Icon from 'react-native-vector-icons/FontAwesome';

import Splash from './src/components/screens/Splash';
import Login from './src/components/screens/Login';
import Home from './src/components/screens/Home';
import Profile from './src/components/screens/Profile';
import CustomDrawer from './src/components/screens/CustomDrawer';

import configureStore from './src/ConfigureStore';
import {Provider} from 'react-redux';
import DeliveryLocation from './src/components/screens/DeliveryLocation';

const store = configureStore();

//TAB ITEMS
const AppTabNavigator1 = createStackNavigator({
  home : {
    screen : Home,
    navigationOptions : ()=>({
      title: 'Home',
    })
  }
},{
  
  defaultNavigationOptions: {
    headerStyle: {
      backgroundColor: 'blue',//'#DD203E',
      borderBottomWidth: 0,
      elevation:0
    },
    headerTintColor: '#fff',
    headerTitleStyle: {
      fontWeight: 'bold',
    },
  } 
})
const AppTabNavigator2 = createStackNavigator({
  pfl : {
    screen : Profile,
    navigationOptions : ()=>({
      title: 'Profile',
    })
  }
},{
  defaultNavigationOptions: {
    headerStyle: {
      backgroundColor: 'blue'//'#DD203E',
    },
    headerTintColor: '#fff',
    headerTitleStyle: {
      fontWeight: 'bold',
    },
  } 
})
//TAB NAVIGATOR
const AppTabNavigator = createBottomTabNavigator({
  tab1 : {
    screen : AppTabNavigator1,
    navigationOptions: () => ({
      tabBarLabel: "Home",
      tabBarIcon: ({ tintColor }) => (
        <Icon name="home" color={tintColor} size={24} />
      )
    })
  },
  tab2 : {
    screen : AppTabNavigator2,
    navigationOptions: () => ({
      tabBarLabel: "Profile",
      tabBarIcon: ({ tintColor }) => (
        <Icon name="user" color={tintColor} size={24} />
      )
    })
  }

},{
  tabBarOptions: {
    activeTintColor: 'white',
    labelStyle: {
      fontSize: 12,
    },
    style: {
      backgroundColor: 'green',
    },
  }
})

//DRAWER 
const AppDrawer = createDrawerNavigator({
  rootTab : {
    screen:AppTabNavigator,
    
  }
},{
  contentComponent:CustomDrawer
})
//ROUTE NAVIGATOR
const AppNavigator = createStackNavigator({
  spls : {
    screen : Splash,
    navigationOptions : ()=>({
      headerBackTitle : null,
      header : null
    })
  },
  log : {
    screen : Login,
    navigationOptions : ()=>({
      headerBackTitle : null,
      title : 'Login',
      header : null,
      gesturesEnabled: false
    })
  },
  deliverylocation : {
    screen:DeliveryLocation,
    navigationOptions : ()=>({
      headerBackTitle : null,
      title : 'Delivery Location',
     // header : null,
     headerLeft:null,
      gesturesEnabled: false
    })
  },
  drawer : {
    screen:AppDrawer,
    navigationOptions : ()=>({
      title: 'Home',
      header : null,
        headerLeft: null,
        gesturesEnabled: false,
        
    })
  }
},{
  headerMode:'screen',
  defaultNavigationOptions: {
    headerStyle: {
      backgroundColor: 'blue'//'#DD203E',
    },
    headerTintColor: '#fff',
    headerTitleStyle: {
      fontWeight: 'bold',
    },
  }  
})
//APP CONTAINER
const AppConatainer = createAppContainer(AppNavigator)

//COMPONENT
export default class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <SafeAreaView style={styles.container} forceInset ={{top : 'never', bottom : 'never'}}>
        <AppConatainer />
      </SafeAreaView>
      </Provider>
     
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Platform.OS == 'ios' ?'#F5FCFF' : 'yellow',
  },
  
});

**Comment by nickjuntilla** [comment]
So my navigator passed parameters would only show up if passed within the same stack in case anyone else was wondering.

**Comment by Humbl-e** [comment]
> Using ```this.props.navigation.navigate("UserInfoScreen", {hello: "world });``` from RegisterStep1, the params get passed.
> 
> Basically navigate to the screen inside the stack rather than to the stack itself.
What if UserInfoStack had another screen say UserDetailScreen, i.e 
```
const UserInfoStack = StackNavigator(
  {
	  UserInfoScreen: UserInfo, 
          UserDetailScreen: UserDetailScreen,
  },
  {
    headerMode: "screen"
  }
);
```
How would i navigate directly to UserDetailScreen? 
Using ```this.props.navigation.navigate("UserDetailsScreen",{});``` from RegisterStep1, crashes the app with error ' props hello is undefined.'

**Comment by GIRIRAJNAGAR768** [comment]
see this....
https://reactnavigation.org/docs/nesting-navigators

it solved my problem.

this.props.navigation.navigate("**stack_Name**", {
 screen:"screen_name_connect_with_**stack_name**",
 params:{
 user:"anything_string_or_object"
}
})
